### DISCUSSION TITLE
Using Assets from other Extensions

### DISCUSSION THREAD

**Discussion by humpalum** [microsoft/vscode-discussions#101] [Extension Development QnA] [answered]
Hey there, I have a question about using Stuff from other extensions /Language Servers:

I am writing an [Extension](https://github.com/humpalum/vscode-sigma) for [Sigma](https://github.com/SigmaHQ/sigma) which is a "language" based on the yaml format. 

I now want to provide stuff like completion and im wondering if i could just use the [YAML extension](https://github.com/redhat-developer/vscode-yaml) as a dependency and use functions from there. I have built a "jsonSchema" for the Sigma format. As far as i understand, YAML uses this to provide completions.  I don't understand their code completely but as far as i understand, they use the language server and provide it the schema.
Is it possible for me to just "piggyback" either the server or some api from YAML?

**Comment by bwateratmsft** [comment, marked as answer]
I think there's two approaches you can do here:

1. Add your schema to https://www.schemastore.org/json (via a PR [here](https://github.com/schemastore/schemastore/)); I think these are automatically picked up by the YAML extension
2. Use the `yaml.schemas` setting provided by the YAML extension. Admittedly, I can't find documentation about how this setting works, but based on the description it sounds like it's for associating schemas to glob file patterns. Your extension could supply a [configuration default](https://code.visualstudio.com/api/references/contribution-points#contributes.configurationDefaults) for `yaml.schemas` that includes your schema and the pattern for Sigma files

Most likely you'll also want to list the YAML extension as an extension dependency of yours (via the `extensionDependencies` property in the [extension manifest](https://code.visualstudio.com/api/references/extension-manifest)), unless you want to make it a soft-dependency.

**Reply by humpalum** [reply]
The problem i have here is that since my extension contributes a new language  the YAML extension is not and thus the Schema checking is not active. Any way of evading that?

**Reply by gjsjohnmurray** [reply]
> is not and thus

Did you mean "is not **activated** and thus"?

**Reply by humpalum** [reply]
Yes, sorry

**Reply by bwateratmsft** [reply]
I could be mistaken but I think that having an extension listed as an `extensionDependency` will cause it to be activated when your extension is activated.

**Comment by gjsjohnmurray** [comment]
In your extension's `activate` method you could activate the YAML one. Something like this:

```ts
const extension = vscode.extensions.getExtension('redhat.vscode-yaml');
if (extension && !extenstion.isActive) {
      await extension.activate();
}
```

**Reply by humpalum** [reply]
Thank you. This helped a bit. I am now able to choose a schema for files even with the "Sigma" language. 
However, as you can see on the screenshot this scheme seems to not be applied.
Could it be, that the YAML extension checks for the language to be YAML and only then applies it checks?

Sadly the YAML extension does not seem to provide commands or an API to manually start these checks. 

Could it be possible to use the yaml language server from my extension?

![image](https://user-images.githubusercontent.com/11175099/185145303-12d807e0-b763-484b-9d93-e6048a789d3a.png)

**Reply by bwateratmsft** [reply]
I would think so, yes. The YAML extension uses the [YAML language server](https://github.com/redhat-developer/yaml-language-server), which is available on [NPM](https://www.npmjs.com/package/yaml-language-server). Your extension could create a language client and have any document selector you'd like.
